### PR TITLE
fix(internal): don't use self-signed JWT with impersonation

### PR DESCRIPTION
**Pull request by codyoss**
The limited audience of the base credential will could cause
failures as it would not be able to comunicate with the iam
credentials api if just default scopes were passed.

### CODE REVIEW

File: internal/creds.go [hunk 1]
```diff
@@ -64,14 +64,18 @@ const (
 
 // credentialsFromJSON returns a google.Credentials based on the input.
 //
-// - If the JSON is a service account and no scopes provided, returns self-signed JWT auth flow
-// - Otherwise, returns OAuth 2.0 flow.
+// - If the JSON is a service account, no user are scopes provided, an audience
+//   is provided, and credentials will not be impersonated executes a
```
Comment (tbpg) [inline internal/creds.go:68]: Missing a word before executes?

File: internal/creds.go [hunk 2]
```diff
@@ -64,14 +64,18 @@ const (
...
+// - Otherwise, executes a stanard OAuth 2.0 flow.
 func credentialsFromJSON(ctx context.Context, data []byte, ds *DialSettings) (*google.Credentials, error) {
 	cred, err := google.CredentialsFromJSON(ctx, data, ds.GetScopes()...)
 	if err != nil {
 		return nil, err
 	}
-	if len(data) > 0 && len(ds.Scopes) == 0 && (ds.DefaultAudience != "" || len(ds.Audiences) > 0) {
+	if len(data) > 0 && len(ds.Scopes) == 0 && 
+	(ds.DefaultAudience != "" || len(ds.Audiences) > 0) && 
+	ds.ImpersonationConfig == nil {
```
Comment (tbpg) [inline internal/creds.go:78]: This is getting pretty complicated. Maybe we should invert the conditions and return early instead? Possibly split into more than one `if` statement.

**Review by tbpg — CHANGES_REQUESTED** [review]

File: internal/creds.go (hunk 1 above)
Comment (codyoss) [reply internal/creds.go]: rephrased.

File: internal/creds.go (hunk 2 above)
Comment (codyoss) [reply internal/creds.go]: Inverted and split up lines to try to improve readability.

**Comment by codyoss** [comment]
@tbpg Could you take a quick look at this when you have a chance.

File: internal/creds.go [hunk 3]
```diff
@@ -64,28 +64,37 @@ const (
 
 // credentialsFromJSON returns a google.Credentials based on the input.
 //
-// - If the JSON is a service account and no scopes provided, returns self-signed JWT auth flow
-// - Otherwise, returns OAuth 2.0 flow.
+// - If the JSON is a service account, no user are scopes provided, an audience
+//   is provided, and credentials will not be impersonated a signed JWT auth
```
Comment (tbpg) [inline internal/creds.go:68]: Missing a word here?

**Review by tbpg — CHANGES_REQUESTED** [review]

**Review by tbpg — APPROVED** [review]

**Pull request merged**